be/src/common/status.cpp

```diff
@@ -4,49 +4,68 @@
 
 #include "common/status.h"
 
+#include <fmt/format.h>
+
 #include "gen_cpp/Status_types.h"  // for TStatus
 #include "gen_cpp/status.pb.h"     // for PStatus
 #include "gutil/strings/fastmem.h" // for memcpy_inlined
 
 namespace starrocks {
 
 // See Status::_state for details.
-static const char g_moved_from_state[8] = {'\x00', '\x00', '\x00', '\x00', TStatusCode::INTERNAL_ERROR,
-                                           '\x00', '\x00', '\x00'};
+static const char g_moved_from_state[5] = {'\x00', '\x00', '\x00', '\x00', TStatusCode::INTERNAL_ERROR};
 
-inline const char* assemble_state(TStatusCode::type code, const Slice& msg, int16_t precise_code, const Slice& msg2) {
+inline const char* assemble_state(TStatusCode::type code, Slice msg, Slice ctx) {
     DCHECK(code != TStatusCode::OK);
 
-    const uint32_t len1 = msg.size;
-    const uint32_t len2 = msg2.size;
-    const uint32_t size = len1 + ((len2 > 0) ? (2 + len2) : 0);
-    auto result = new char[size + 7];
-    memcpy(result, &size, sizeof(size));
+    msg.size = std::min<size_t>(msg.size, std::numeric_limits<uint16_t>::max());
+    ctx.size = std::min<size_t>(ctx.size, std::numeric_limits<uint16_t>::max());
+
+    const uint16_t len1 = msg.size;
+    const uint16_t len2 = ctx.size;
+    const uint32_t size = static_cast<uint32_t>(len1) + len2;
+    auto result = new char[size + 5];
+    memcpy(result, &len1, sizeof(len1));
+    memcpy(result + 2, &len2, sizeof(len2));
     result[4] = static_cast<char>(code);
-    memcpy(result + 5, &precise_code, sizeof(precise_code));
-    memcpy(result + 7, msg.data, len1);
-    if (len2 > 0) {
-        result[7 + len1] = ':';
-        result[8 + len1] = ' ';
-        memcpy(result + 9 + len1, msg2.data, len2);
-    }
+    strings::memcpy_inlined(result + 5, msg.data, len1);
+    strings::memcpy_inlined(result + 5 + len1, ctx.data, len2);
+    return result;
+}
+
+const char* Status::copy_state(const char* state) const {
+    uint16_t len1;
+    uint16_t len2;
+    strings::memcpy_inlined(&len1, state, sizeof(len1));
+    strings::memcpy_inlined(&len2, state + sizeof(len1), sizeof(len2));
+    uint32_t length = static_cast<uint32_t>(len1) + len2 + 5;
+    auto result = new char[length];
+    strings::memcpy_inlined(result, state, length);
     return result;
 }
 
-const char* Status::copy_state(const char* state) {
-    uint32_t size;
-    strings::memcpy_inlined(&size, state, sizeof(size));
-    auto result = new char[size + 7];
-    strings::memcpy_inlined(result, state, size + 7);
+const char* Status::copy_state_with_extra_ctx(const char* state, Slice ctx) const {
+    uint16_t len1;
+    uint16_t len2;
+    strings::memcpy_inlined(&len1, state, sizeof(len1));
+    strings::memcpy_inlined(&len2, state + sizeof(len1), sizeof(len2));
+    uint32_t old_length = static_cast<uint32_t>(len1) + len2 + 5;
+    ctx.size = std::min<size_t>(ctx.size, std::numeric_limits<uint16_t>::max() - old_length);
```

Comment:
When std::numeric_limits<uint16_t>::max() is less than old_length?

Reply:
Good point, `std::numeric_limits<uint16_t>::max() - old_length` is incorrect, should be `std::numeric_limits<uint16_t>::max() - len2` here.

Reply:
fixed